populator-machinery/controller.go

```diff
@@ -527,13 +557,53 @@ func (c *controller) syncPvc(ctx context.Context, key, pvcNamespace, pvcName str
 	}
 
 	// Look for the populator pod
-	podName := fmt.Sprintf("%s-%s", populatorPodPrefix, pvc.UID)
-	c.addNotification(key, "pod", c.populatorNamespace, podName)
 	var pod *corev1.Pod
-	pod, err = c.podLister.Pods(c.populatorNamespace).Get(podName)
-	if err != nil {
-		if !errors.IsNotFound(err) {
-			return err
+	podName := fmt.Sprintf("%s-%s", populatorPodPrefix, pvc.UID)
+	if !c.useProviderImpl {
+		c.addNotification(key, "pod", c.populatorNamespace, podName)
+		pod, err = c.podLister.Pods(c.populatorNamespace).Get(podName)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				return err
+			}
+		}
+	}
+
+	// If populate data with cloud provider implementation, and orginal StorageClass's VolumeBindingMode is VolumeBindingWaitForFirstConsumer,
+	// create a StorageClass with VolumeBindingImmediate for pvcPrime
+	storageClassPrimeName := *pvc.Spec.StorageClassName
+	if c.useProviderImpl && storageClass.VolumeBindingMode != nil && storagev1.VolumeBindingWaitForFirstConsumer == *storageClass.VolumeBindingMode {
+		storageClassPrimeName = fmt.Sprintf("%s-%s", populatorStorageClassPrefix, *pvc.Spec.StorageClassName)
+		scPrime, err := c.scLister.Get(storageClassPrimeName)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				return err
+			}
+			c.addNotification(key, "sc", "", storageClassPrimeName)
+			bm := storagev1.VolumeBindingImmediate
+			scPrime = &storagev1.StorageClass{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:        storageClassPrimeName,
+					Annotations: map[string]string{},
+				},
+				Provisioner:          storageClass.Provisioner,
+				Parameters:           storageClass.Parameters,
+				ReclaimPolicy:        storageClass.ReclaimPolicy,
+				MountOptions:         storageClass.MountOptions,
+				AllowVolumeExpansion: storageClass.AllowVolumeExpansion,
+				VolumeBindingMode:    &bm,
+				AllowedTopologies:    storageClass.AllowedTopologies,
+			}
+			scPrime.ObjectMeta.Annotations[annStorageClassPrime] = storageClassPrimeName
```

Comment:
What is the purpose of this annotation? It seems to be duplicating the name that is available in the objectMeta.Name

Reply:
I'm using the key to indicate that the storageclass is created by the volume-populator https://github.com/kubernetes-csi/lib-volume-populator/pull/133#discussion_r1490208385
Is there any example of how to annotate a resource created by a specific component?

Reply:
I think a finalizer to ensure the user does not delete the SCPrime, may be sufficient, given that we can map SCorig back to SCprime. i.e we may need to do this mapping when we watch the SCorig and delete SCprime on SCorig delete event

Thoughts?


> Is there any example of how to annotate a resource created by a specific component?
The annotation can be any KV pair. But we should keep the prefix of the key consistent to how it's done in https://github.com/kubernetes-csi/lib-volume-populator/blob/master/populator-machinery/controller.go#L167.  (If we use the finalizer that may be sufficient given that we can map SC back to SC Prime name. i.e if we watch the SCorig and delete SC prime on SCorig delete event)

Reply:
Make sense, I think this should be implement with the StorageClass prime auto cleanup, otherwise there will be finalizer added without removal.

Reply:
I add a TODO, and remove the annotation.